src/provider/provider_os_memory.c

```diff
@@ -583,21 +672,162 @@ static const char *os_get_name(void *provider) {
 
 static umf_result_t os_allocation_split(void *provider, void *ptr,
```

Comment:
I think this function is supposed to be thread-safe. Currently, we take a lock in the tracking provider and this function would be called under that lock so everything is fine. We might want to add some kind of assert here to make sure that tracking is enabled though.

Reply:
> I think this function is supposed to be thread-safe. Currently, we take a lock in the tracking provider and this function would be called under that lock so everything is fine.

If this function is supposed to be thread-safe it does not matter if it is called under the lock. Or I did not get your point. 

> We might want to add some kind of assert here to make sure that tracking is enabled though.

I think split/merge functionality is independent of memory tracking capabilities, right?

Anyway, I think in general @igchor raised a good point, we should specify the requirements for memory provider interface implementations. 
In this particular case, do we assume that split/merge can be called concurrently for the same memory region? I think no but maybe I miss something. If split/merge is called concurrently for different memory regions I think it is fine.

I would say the following:

- The implementation of these functions should be thread-safe.
- It should not be called concurrently with the same pointer.

Reply:
I think we do allow calling split/merge on the same memory region concurrently right now - that's why we have a mutex in memory tracker.

Reply:
An assert added.
Done.

Reply:
> An assert added.

But what if the tracker is disabled? Do we want to support split/merge in that case?

> I think we do allow calling split/merge on the same memory region concurrently right now

Do we have such scenarios? For me, it is hard to imagine. Maybe it is better to specify that "It should **not** be called concurrently with the same pointer."

Reply:
Yeah, we all have to agree upon one version first. I removed this assert and left a comment only.

Reply:
Done